File: implementations/java/src/main/java/com/webweavex/WebWeaveX.java
package com.webweavex;

import java.util.*;
import java.time.Instant;

public class WebWeaveX {
    private Pipeline pipeline;
    
    public WebWeaveX() {
        this.pipeline = new Pipeline();
    }
    
    public Map<String, Object> extract(String textOrHtml) {
        try {
            if (textOrHtml == null || textOrHtml.isEmpty()) {
                return pipeline.extractFromText("");
            }
            
            if (textOrHtml.trim().startsWith("<") && textOrHtml.toLowerCase().contains("</html>")) {
                return pipeline.extractFromText(textOrHtml);
            }
            
            return pipeline.extractFromText(textOrHtml);
        } catch (Exception e) {
            Map<String, Object> errorResult = new LinkedHashMap<>();
            Map<String, String> meta = new LinkedHashMap<>();
            meta.put("title", "");
            meta.put("url", "");
            Map<String, String> content = new LinkedHashMap<>();
            content.put("text", "");
            
            errorResult.put("meta", meta);
            errorResult.put("content", content);
            errorResult.put("chunks", new ArrayList<>());
            errorResult.put("entities", new ArrayList<>());
            errorResult.put("relations", new ArrayList<>());
            
            Map<String, Object> graph = new LinkedHashMap<>();
            graph.put("nodes", new ArrayList<>());
            graph.put("edges", new ArrayList<>());
            errorResult.put("graph", graph);
            
            Map<String, Object> insights = new LinkedHashMap<>();
            insights.put("entity_counts", new LinkedHashMap<>());
            insights.put("stats", new LinkedHashMap<>());
            insights.put("top_entities", new ArrayList<>());
            errorResult.put("insights", insights);
            
            return errorResult;
        }
    }
    
    public Map<String, Object> extractAgent(String text) {
        try {
            Map<String, Object> result = extract(text);
            return extractAgentFromResult(result);
        } catch (Exception e) {
            Map<String, Object> agentResult = new LinkedHashMap<>();
            agentResult.put("task", "web_analysis");
            agentResult.put("input", text != null ? text.substring(0, Math.min(500, text.length())) : "");
            agentResult.put("output", new LinkedHashMap<>());
            agentResult.put("summary", "Error: " + e.getMessage());
            agentResult.put("actions", new ArrayList<>());
            agentResult.put("confidence", 0.0);
            return agentResult;
        }
    }
    
    @SuppressWarnings("unchecked")
    private Map<String, Object> extractAgentFromResult(Map<String, Object> result) {
        Map<String, Object> agentResult = new LinkedHashMap<>();
        agentResult.put("task", "web_analysis");
        
        Map<String, Object> content = (Map<String, Object>) result.get("content");
        String text = content != null ? (String) content.getOrDefault("text", "") : "";
        agentResult.put("input", text.length() > 500 ? text.substring(0, 500) : text);
        
        List<Map<String, String>> entities = (List<Map<String, String>>) result.get("entities");
        List<Map<String, String>> relations = (List<Map<String, String>>) result.get("relations");
        
        if (entities == null) entities = new ArrayList<>();
        if (relations == null) relations = new ArrayList<>();
        
        agentResult.put("output", result);
        
        if (entities.isEmpty()) {
            agentResult.put("summary", "No entities extracted from input text.");
        } else {
            agentResult.put("summary", String.format("Extracted %d entities from text.", entities.size()));
        }
        
        List<String> actions = new ArrayList<>();
        Set<String> types = new HashSet<>();
        for (Map<String, String> e : entities) {
            types.add(e.get("type"));
        }
        if (types.contains("url")) actions.add("crawl");
        if (types.contains("email")) actions.add("contact");
        if (types.contains("phone")) actions.add("call");
        if (entities.size() > 5) actions.add("extract_more");
        if (actions.isEmpty()) actions.add("analyze");
        agentResult.put("actions", actions);
        
        double confidence = text.length() > 0 
            ? Math.min((entities.size() + relations.size() * 0.5) / Math.max(text.length(), 1) * 10, 1.0) 
            : 0.0;
        agentResult.put("confidence", Math.round(confidence * 100.0) / 100.0);
        
        return agentResult;
    }
    
    @SuppressWarnings("unchecked")
    public Map<String, Object> toMemoryBlock(Map<String, Object> result) {
        Map<String, Object> memory = new LinkedHashMap<>();
        memory.put("type", "webweavex_memory");
        memory.put("entities", result.getOrDefault("entities", new ArrayList<>()));
        memory.put("relations", result.getOrDefault("relations", new ArrayList<>()));
        memory.put("graph", result.getOrDefault("graph", createEmptyGraph()));
        memory.put("timestamp", Instant.now().toString());
        memory.put("source", "webweavex");
        return memory;
    }
    
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> toRagChunks(Map<String, Object> result) {
        List<Map<String, Object>> ragChunks = new ArrayList<>();
        List<Map<String, Object>> chunks = (List<Map<String, Object>>) result.get("chunks");
        List<Map<String, String>> entities = (List<Map<String, String>>) result.get("entities");
        List<Map<String, String>> relations = (List<Map<String, String>>) result.get("relations");
        
        if (chunks == null) chunks = new ArrayList<>();
        if (entities == null) entities = new ArrayList<>();
        if (relations == null) relations = new ArrayList<>();
        
        for (Map<String, Object> chunk : chunks) {
            Map<String, Object> ragChunk = new LinkedHashMap<>();
            ragChunk.put("text", chunk.get("text"));
            
            Map<String, Object> metadata = new LinkedHashMap<>();
            metadata.put("entities", entities);
            metadata.put("relations", relations.subList(0, Math.min(5, relations.size())));
            metadata.put("source", "webweavex");
            ragChunk.put("metadata", metadata);
            
            ragChunks.add(ragChunk);
        }
        
        return ragChunks;
    }
    
    public Iterable<String> extractStream(String text) {
        return () -> new Iterator<String>() {
            private int stage = 0;
            private final String[] stages = {"cleaning", "chunking", "entities", "relations", "graph", "insights"};
            
            public boolean hasNext() {
                return stage < stages.length;
            }
            
            public String next() {
                return stages[stage++];
            }
        };
    }
    
    public String prettyPrint(Map<String, Object> result) {
        StringBuilder sb = new StringBuilder();
        sb.append("==================================================\n");
        sb.append("WebWeaveX Analysis\n");
        sb.append("==================================================\n\n");
        sb.append("ENTITY SUMMARY:\n");
        sb.append("------------------------------\n");
        
        Map<String, Object> insights = (Map<String, Object>) result.get("insights");
        if (insights != null) {
            Map<String, Integer> entityCounts = (Map<String, Integer>) insights.get("entity_counts");
            if (entityCounts != null) {
                for (Map.Entry<String, Integer> entry : entityCounts.entrySet()) {
                    sb.append(String.format("  %s: %d\n", entry.getKey(), entry.getValue()));
                }
            }
            
            Map<String, Object> stats = (Map<String, Object>) insights.get("stats");
            if (stats != null) {
                sb.append("\nSTATISTICS:\n");
                sb.append("------------------------------\n");
                sb.append(String.format("  Total Entities: %s\n", stats.getOrDefault("total_entities", 0)));
                sb.append(String.format("  Unique Entities: %s\n", stats.getOrDefault("unique_entities", 0)));
                sb.append(String.format("  Entity Types: %s\n", stats.getOrDefault("entity_types", 0)));
                sb.append(String.format("  Total Relations: %s\n", stats.getOrDefault("total_relations", 0)));
                sb.append(String.format("  Total Chunks: %s\n", stats.getOrDefault("total_chunks", 0)));
                sb.append(String.format("  Text Length: %s\n", stats.getOrDefault("text_length", 0)));
                sb.append(String.format("  Word Count: %s\n", stats.getOrDefault("word_count", 0)));
            }
        }
        
        sb.append("\n==================================================\n");
        return sb.toString();
    }
    
    public static Map<String, Object> getToolSchema() {
        Map<String, Object> schema = new LinkedHashMap<>();
        schema.put("name", "webweavex_extract");
        schema.put("description", "Extract structured intelligence from text");
        
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("type", "object");
        
        Map<String, Object> properties = new LinkedHashMap<>();
        Map<String, String> inputProp = new LinkedHashMap<>();
        inputProp.put("type", "string");
        properties.put("input", inputProp);
        params.put("properties", properties);
        
        List<String> required = new ArrayList<>();
        required.add("input");
        params.put("required", required);
        
        schema.put("parameters", params);
        return schema;
    }
    
    public static List<Map<String, Object>> getAllTools() {
        List<Map<String, Object>> tools = new ArrayList<>();
        tools.add(getToolSchema());
        
        Map<String, Object> entitiesTool = new LinkedHashMap<>();
        entitiesTool.put("name", "webweavex_entities");
        entitiesTool.put("description", "Extract only entities from text");
        
        Map<String, Object> entitiesParams = new LinkedHashMap<>();
        entitiesParams.put("type", "object");
        Map<String, Object> entitiesProps = new LinkedHashMap<>();
        entitiesProps.put("input", Map.of("type", "string"));
        entitiesParams.put("properties", entitiesProps);
        entitiesParams.put("required", List.of("input"));
        entitiesTool.put("parameters", entitiesParams);
        tools.add(entitiesTool);
        
        return tools;
    }
    
    public static List<String> getCapabilities() {
        return Arrays.asList(
            "extract", "entities", "graph", "rag", 
            "agent_mode", "memory_export", "streaming"
        );
    }
    
    private Map<String, Object> createEmptyGraph() {
        Map<String, Object> graph = new LinkedHashMap<>();
        graph.put("nodes", new ArrayList<>());
        graph.put("edges", new ArrayList<>());
        return graph;
    }
    
    public Pipeline getPipeline() {
        return pipeline;
    }
    
    public static void main(String[] args) {
        WebWeaveX wx = new WebWeaveX();
        Map<String, Object> result = wx.extract("Contact test@example.com 555-1234");
        System.out.println("WebWeaveX Java initialized successfully");
    }
}

File: implementations/kotlin/src/main/kotlin/com/webweavex/WebWeaveX.kt
package com.webweavex

import java.time.Instant

class WebWeaveX {
    private val pipeline = Pipeline()

    fun extract(textOrHtml: String): Map<String, Any> {
        return try {
            if (textOrHtml.isEmpty()) {
                pipeline.extractFromText("")
            } else if (textOrHtml.trim().startsWith("<") && textOrHtml.lowercase().contains("</html>")) {
                pipeline.extractFromText(textOrHtml)
            } else {
                pipeline.extractFromText(textOrHtml)
            }
        } catch (e: Exception) {
            createErrorResult()
        }
    }

    fun clean(text: String): String {
        return pipeline.extractFromText(text)["content"]?.let { 
            (it as? Map<*, *>)?.get("text") as? String ?: ""
        } ?: ""
    }

    fun chunk(text: String): List<Chunk> {
        return pipeline.extractFromText(text)["chunks"] as? List<Chunk> ?: emptyList()
    }

    fun entities(text: String): List<Entity> {
        return pipeline.extractFromText(text)["entities"] as? List<Entity> ?: emptyList()
    }

    fun graph(text: String): GraphData {
        return pipeline.extractFromText(text)["graph"] as? GraphData ?: GraphData(emptyList(), emptyList())
    }

    fun extractAgent(text: String): Map<String, Any> {
        return try {
            val result = extract(text)
            extractAgentFromResult(result, text)
        } catch (e: Exception) {
            linkedMapOf(
                "task" to "web_analysis",
                "input" to (if (text.length > 500) text.substring(0, 500) else text),
                "output" to linkedMapOf<String, Any>(),
                "summary" to "Error: ${e.message}",
                "actions" to emptyList<String>(),
                "confidence" to 0.0
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun extractAgentFromResult(result: Map<String, Any>, inputText: String): Map<String, Any> {
        val content = result["content"] as? Map<*, *> ?: emptyMap<Any, Any>()
        val text = content["text"] as? String ?: ""
        val entities = result["entities"] as? List<Map<String, String>> ?: emptyList()
        val relations = result["relations"] as? List<Map<String, String>> ?: emptyList()

        val summary = if (entities.isEmpty()) {
            "No entities extracted from input text."
        } else {
            "Extracted ${entities.size} entities from text."
        }

        val actions = mutableListOf<String>()
        val types = entities.mapNotNull { it["type"] }.toSet()
        if ("url" in types) actions.add("crawl")
        if ("email" in types) actions.add("contact")
        if ("phone" in types) actions.add("call")
        if (entities.size > 5) actions.add("extract_more")
        if (actions.isEmpty()) actions.add("analyze")

        val confidence = if (text.isNotEmpty()) {
            minOf((entities.size + relations.size * 0.5) / text.length.toDouble() * 10, 1.0)
        } else {
            0.0
        }

        return linkedMapOf(
            "task" to "web_analysis",
            "input" to (if (text.length > 500) text.substring(0, 500) else text),
            "output" to result,
            "summary" to summary,
            "actions" to actions,
            "confidence" to (confidence * 100).toInt() / 100.0
        )
    }

    fun toMemoryBlock(result: Map<String, Any>): Map<String, Any> {
        return try {
            linkedMapOf(
                "type" to "webweavex_memory",
                "entities" to (result["entities"] ?: emptyList<Map<String, Any>>()),
                "relations" to (result["relations"] ?: emptyList<Map<String, Any>>()),
                "graph" to (result["graph"] ?: createEmptyGraph()),
                "timestamp" to Instant.now().toString(),
                "source" to "webweavex"
            )
        } catch (e: Exception) {
            linkedMapOf(
                "type" to "webweavex_memory",
                "entities" to emptyList<Map<String, Any>>(),
                "relations" to emptyList<Map<String, Any>>(),
                "graph" to createEmptyGraph(),
                "timestamp" to Instant.now().toString(),
                "source" to "webweavex"
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun toRagChunks(result: Map<String, Any>): List<Map<String, Any>> {
        return try {
            val chunks = result["chunks"] as? List<Map<String, Any>> ?: emptyList()
            val entities = result["entities"] as? List<Map<String, String>> ?: emptyList()
            val relations = result["relations"] as? List<Map<String, String>> ?: emptyList()

            chunks.map { chunk ->
                linkedMapOf(
                    "text" to (chunk["text"] ?: ""),
                    "metadata" to linkedMapOf(
                        "entities" to entities,
                        "relations" to relations.take(5),
                        "source" to "webweavex"
                    )
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun extractStream(text: String): Sequence<String> {
        return sequenceOf("cleaning", "chunking", "entities", "relations", "graph", "insights")
    }

    fun prettyPrint(result: Map<String, Any>): String {
        val sb = StringBuilder()
        sb.appendLine("==================================================")
        sb.appendLine("WebWeaveX Analysis")
        sb.appendLine("==================================================")
        sb.appendLine()
        sb.appendLine("ENTITY SUMMARY:")
        sb.appendLine("------------------------------")

        val insights = result["insights"] as? Map<*, *>
        val entityCounts = insights?.get("entity_counts") as? Map<String, Int>
        entityCounts?.forEach { (key, count) ->
            sb.appendLine("  $key: $count")
        }

        val stats = insights?.get("stats") as? Map<String, Any>
        if (stats != null) {
            sb.appendLine()
            sb.appendLine("STATISTICS:")
            sb.appendLine("------------------------------")
            sb.appendLine("  Total Entities: ${stats["total_entities"] ?: 0}")
            sb.appendLine("  Unique Entities: ${stats["unique_entities"] ?: 0}")
            sb.appendLine("  Entity Types: ${stats["entity_types"] ?: 0}")
            sb.appendLine("  Total Relations: ${stats["total_relations"] ?: 0}")
            sb.appendLine("  Total Chunks: ${stats["total_chunks"] ?: 0}")
            sb.appendLine("  Text Length: ${stats["text_length"] ?: 0}")
            sb.appendLine("  Word Count: ${stats["word_count"] ?: 0}")
        }

        sb.appendLine()
        sb.appendLine("==================================================")
        return sb.toString()
    }

    companion object {
        fun getToolSchema(): Map<String, Any> {
            return linkedMapOf(
                "name" to "webweavex_extract",
                "description" to "Extract structured intelligence from text",
                "parameters" to linkedMapOf(
                    "type" to "object",
                    "properties" to linkedMapOf(
                        "input" to linkedMapOf("type" to "string")
                    ),
                    "required" to listOf("input")
                )
            )
        }

        fun getAllTools(): List<Map<String, Any>> {
            return listOf(
                getToolSchema(),
                linkedMapOf(
                    "name" to "webweavex_entities",
                    "description" to "Extract only entities from text",
                    "parameters" to linkedMapOf(
                        "type" to "object",
                        "properties" to linkedMapOf(
                            "input" to linkedMapOf("type" to "string")
                        ),
                        "required" to listOf("input")
                    )
                )
            )
        }

        fun getCapabilities(): List<String> {
            return listOf(
                "extract", "entities", "graph", "rag",
                "agent_mode", "memory_export", "streaming"
            )
        }

        private fun createEmptyGraph(): Map<String, Any> {
            return linkedMapOf(
                "nodes" to emptyList<Map<String, Any>>(),
                "edges" to emptyList<Map<String, Any>>()
            )
        }
    }

    private fun createErrorResult(): Map<String, Any> {
        return linkedMapOf(
            "meta" to linkedMapOf("title" to "", "url" to ""),
            "content" to linkedMapOf("text" to ""),
            "chunks" to emptyList<Map<String, Any>>(),
            "entities" to emptyList<Map<String, Any>>(),
            "relations" to emptyList<Map<String, Any>>(),
            "graph" to createEmptyGraph(),
            "insights" to linkedMapOf(
                "entity_counts" to emptyMap<String, Int>(),
                "stats" to emptyMap<String, Any>(),
                "top_entities" to emptyList<Map<String, Any>>()
            )
        )
    }
}
